include/sparrow/arrow_array_schema_proxy.hpp

```diff
@@ -0,0 +1,467 @@
+            case data_type::DOUBLE:
+            case data_type::HALF_FLOAT:
+            case data_type::TIMESTAMP:
+            case data_type::FIXED_SIZE_BINARY:
+            case data_type::DECIMAL:
+            case data_type::FIXED_WIDTH_BINARY:
+                return {buffer_type::VALIDITY, buffer_type::DATA};
+            case data_type::STRING:
+                return {buffer_type::VALIDITY, buffer_type::OFFSETS_32BIT, buffer_type::DATA};
+            case data_type::LIST:
+                return {buffer_type::VALIDITY, buffer_type::OFFSETS_32BIT};
+            case data_type::LARGE_LIST:
+                return {buffer_type::VALIDITY, buffer_type::OFFSETS_64BIT};
+            case data_type::LIST_VIEW:
+                return {buffer_type::VALIDITY, buffer_type::OFFSETS_32BIT, buffer_type::SIZES_32BIT};
+            case data_type::LARGE_LIST_VIEW:
+                return {buffer_type::VALIDITY, buffer_type::OFFSETS_64BIT, buffer_type::SIZES_64BIT};
+            case data_type::FIXED_SIZED_LIST:
+            case data_type::STRUCT:
+                return {buffer_type::VALIDITY};
+            case data_type::SPARSE_UNION:
+                return {buffer_type::TYPE_IDS};
+            case data_type::DENSE_UNION:
+                return {buffer_type::TYPE_IDS, buffer_type::OFFSETS_32BIT};
+            case data_type::NA:
+            case data_type::MAP:
+            case data_type::RUN_ENCODED:
+                return {};
+        }
+        mpl::unreachable();
+    }
+
+    constexpr std::size_t get_offset_size(data_type data_type, int64_t length)
+    {
+        switch (data_type)
+        {
+            case data_type::STRING:
+            case data_type::LIST:
+            case data_type::LARGE_LIST:
+                return static_cast<std::size_t>(length) + 1;
+            case data_type::LIST_VIEW:
+            case data_type::LARGE_LIST_VIEW:
+            case data_type::DENSE_UNION:
+                return static_cast<std::size_t>(length);
+            default:
+                throw std::runtime_error("Unsupported data type");
+        }
+    }
+
+    class arrow_proxy
+    {
+    public:
+
+        explicit arrow_proxy(ArrowArray&& array, ArrowSchema&& schema);
+        explicit arrow_proxy(ArrowArray* array, ArrowSchema* schema);
+
+        [[nodiscard]] std::string_view format() const;
+        [[nodiscard]] std::string_view name() const;
+        [[nodiscard]] std::string_view metadata() const;
+        [[nodiscard]] std::vector<ArrowFlag> flags() const;
+        [[nodiscard]] int64_t length() const;
+        [[nodiscard]] int64_t null_count() const;
+        [[nodiscard]] int64_t offset() const;
+        [[nodiscard]] int64_t n_buffers() const;
+        [[nodiscard]] int64_t n_children() const;
+        [[nodiscard]] std::vector<sparrow::buffer_view<uint8_t>> buffers() const;
+        [[nodiscard]] std::vector<arrow_proxy> children() const;
+        [[nodiscard]] std::optional<arrow_proxy> dictionary() const;
```

Comment:
In that case it might be better to handle that in the array implementation classes. We can probably factorize out the creation of the bitmap view in an intermediate class rather than using an std::optional for something known at build time.